packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,222 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import autobind from 'autobind-decorator'
+import themed from 'react-themed'
+import { EVENTS } from './utils/const'
+import { setupEvents, removeEvent } from './utils/mapEventHelpers'
+
+@themed(['Gmap_FreeDrawTootip'], { pure: true })
+export default class FreeDrawLayer extends PureComponent {
+  static propTypes = {
+    map: PropTypes.object,
+    onMapDrawStart: PropTypes.func,
+    onMapDrawEnd: PropTypes.func,
+    theme: PropTypes.object,
+    mapControls: PropTypes.object,
+    shapes: PropTypes.object,
+    multipleShapes: PropTypes.bool,
+    disabled: PropTypes.bool,
+    dataStyle: PropTypes.object,
+  }
+
+  static defaultProps = {
+    theme: {},
+    map: {},
+    multipleShapes: false,
+    disabled: false,
+    shapes: {},
+    dataStyle: {},
+  }
+
+  constructor(props) {
+    super(props)
+    this.state = {
+      drawing: false,
+      polygons: (Object.keys(this.props.shapes).length > 0) ? [this.addShapeToGMaps(
+        this.formatLongLatToGmapsCoordinates(
+          this.props.shapes
+        )[0]
+      )] : [],
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.shapes !== prevProps.shapes) {
+      this.clearAllShapes()
+      this.addShape(
+        this.formatLongLatToGmapsCoordinates(
+          this.props.shapes
+        )[0]
+      )
+    }
+  }
+
+  getPolygonCoords(poly) {
+    const len = poly.getPath().getLength()
+    const shape = []
+
+    for (let i = 0; i < len; i += 1) {
+      shape.push(
+        poly.getPath()
+          .getAt(i)
+          .toUrlValue(5)
+          .split(',')
+          .map(Number)
+          .reverse()
```

Comment:
Instead of this string manipulation you could do...
```
shape.push([poly.getPath().getAt(i).lng(), poly.getPath().getAt(i).lat()])
```
Also you might consider calling `getPath` just once before the loop and assigning it to a local.

Reply:
I think what I proposed is more direct/clear (and probably a little faster fwiw).

Reply:
I agree - I took a look at the Google Maps API and the polygon object does return the lat() lng() methods so it's easier to manipulate.